### ISSUE TITLE
Batched GESV or similar on GPU machines

### ISSUE THREAD

**Issue by wjhorne** [kokkos/kokkos-kernels#1349]
Hi,

I have a workload where I need to solve a large list of small linear systems and have had trouble finding a good solution using Kokkos kernels to deal with it. I tried a batched LU solver solution, but ran into accuracy issues which I believe are related to the lack of pivoting. GESV works well but appears to be host only unless I use MAGMA. I do have access to third party BLAS implementations, such as cuBLAS, but I don't have ready access to MAGMA everywhere so it isn't a great option.

Is there a good method available using Kokkos kernels to tackle this workload? I was thinking that something like a batched GESV might be ideal

**Comment by kliegeois** [comment]
Hello,
What are the sizes of your system and how many of them do you have?
Are the matrices dense or sparse?
Are the matrices SPD?

There are implementations of batched iterative solver at the team level (on the device), I don't know if it could be useful.

**Comment by wjhorne** [comment]
They vary from 4x4 to 100x100 depending on the case. At most I expect 10k or so systems. They are dense and they are symmetric but not positive definite. Would the iterative solver be appropriate for such a system?

I am also working on doing a more rigorous sensitivity study of the TinyValue for the non pivoting LU-solver solution to see if there is a value that would be acceptable.

**Comment by kliegeois** [comment]
For a given simulation are they all of the same size or can there be some difference in the sizes within a batched system?

The batched CG cannot work as your matrices are not SPD but the batched GMRES could be used. Whether it would be efficient or not would depend on the spectra of your matrices.

**Comment by wjhorne** [comment]
The sizes vary within a batch. I generally have used routines like SerialLU within a parallel for over each linear system.  I have now verified that there really isn't a TinyValue that will work for my systems. I am seeing errors up to O(1e-3) with it vs 1e-11 using gesv with pivoting on the host. This is the best I have found over a large span of tested values.

Is there a version of the GMRES routine that can be called within a device kernel? Something akin to SerialLU?

**Comment by kliegeois** [comment]
What is your outermost Kokkos parallel for loop around the batched solve?
Is it a team parallel for loop?

If yes, how many system(s) per team do you have?
If you have 1 system per team, you could use this:

https://github.com/kokkos/kokkos-kernels/blob/develop/src/batched/sparse/KokkosBatched_GMRES.hpp

If you have more than 1 system per team, they need to share the same dimensions.

If you do not have a team parallel for, there is no way to use the batched GMRES for now.

**Comment by wjhorne** [comment]
I am currently using something close to the following pattern. AA and BB are the full batched system, N is the number of systems, and n is the local system size. I have not been using team parallel for but can switch over if it allows for a better solution via GMRES.

Kokkos::parallel_for(N, KOKKOS_LAMBDA(const int i) {

  auto A = Kokkos::subview(AA, i, Kokkos::make_pair(0, n), Kokkos::make_pair(0, n)); 
  auto B = Kokkos::subview(BB, i, Kokkos::make_pair(0,n));  
            
  KokkosBatched::SerialLU<KokkosBatched::Algo::LU::Unblocked>::invoke(A, TINY_VALUE);
  KokkosBatched::SerialSolveLU<KokkosBatched::Trans::NoTranspose, 
                                                        KokkosBatched::Algo::SolveLU::Unblocked>::invoke(A, B);

});

**Comment by kliegeois** [comment]
I think that it could work using a team parallel for, however, the current GMRES implementation only supports sparse matrices for now.
I can, in the near future, provide an interface for a dense matrix.

**Comment by wjhorne** [comment]
Would it be possible to instead go for a direct solve with partial pivoting instead for the future (near or further)? Or is that much less tractable than a GMRES dense matrix? I can also provide you a simple example matrix if it would be helpful.

For now I have put together a solution that will use tiny_value on the device or a host GESV call that I'll plan to change in the future depending on what capability is present.

**Comment by kliegeois** [comment]
I have to discuss with the batched dense developers but it might be possible.
If I may ask, on which application code do you work?

**Comment by wjhorne** [comment]
This would all be for SPARC. I should also clarify that I am not trying to offload work on the kokkos kernels team if it is out of scope or otherwise. I just think that if effort is going to be put towards this then it might be better placed on adding pivoting rather than GMRES

**Comment by kliegeois** [comment]
@wjhorne Can I close this issue or is there something else that should be done?

**Comment by wjhorne** [comment]
This can be closed. I will also respond to your recent message soon internally.
